### ISSUE TITLE
/api/query/exp fails when using some types of nested expressions

### ISSUE THREAD

**Issue by johann8384** [OpenTSDB/opentsdb#834]
Will work:

```
    "expressions": [
		{ "id": "e1", "expr": "m1" },
		{ "id": "e2", "expr": "m2" },
		{ "id": "e3", "expr": "m3" },
		{ "id": "e12","expr": "e1+e2+e3" }
    ]
```
```
    "expressions": [
		{ "id": "e1", "expr": "m1" },
		{ "id": "e2", "expr": "m2" },
		{ "id": "e3", "expr": "m3" },
		{ "id": "e12","expr": "m1+m2" }
    ]
```
Will fail:

```
    "expressions": [
		{ "id": "e1", "expr": "m1" },
		{ "id": "e2", "expr": "m2" },
		{ "id": "e3", "expr": "m3" },
		{ "id": "e12","expr": "e1" }
    ]
```
```
    "expressions": [
		{ "id": "e1", "expr": "m1" },
		{ "id": "e2", "expr": "m2" },
		{ "id": "e3", "expr": "m3" },
    	{ "id": "e12","expr": "e1+e3" }
    ]
```

**Comment by johann8384** [comment]
Complete Query:
http://localhost:4242/api/query/exp
```
{
    "time": {
        "start": "1h-ago",
        "aggregator": "sum"
    },
    "filters": [
      { "id": "f1", "tags": [{
      "type":"wildcard",
      "tagk":"status",
      "filter":"*",
      "groupBy":true
    }] }
    ],
    "metrics": [
        { "id": "m1", "metric": "interface.p2p0.if_packets.tx", "filter": "f1" },
		{ "id": "m2", "metric": "interface.p2p0.if_packets.tx", "filter": "f1" },
      	{ "id": "m3", "metric": "interface.bridge0.if_octets.rx", "filter": "f1" }
    ],
    "expressions": [
		{ "id": "e1", "expr": "m1" },
		{ "id": "e2", "expr": "m2" },
		{ "id": "e3", "expr": "m3" },
    	{ "id": "e12","expr": "e1+e2+e3" }
    ]
}
```

**Comment by johann8384** [comment]
Output from the error:

```
2016-07-12 19:03:20,896 INFO  [OpenTSDB I/O Worker #3] ConnectionManager: [id: 0x3518b469, /0:0:0:0:0:0:0:1:62207 => /0:0:0:0:0:0:0:1:4242] BOUND: /0:0:0:0:0:0:0:1:4242
2016-07-12 19:03:20,896 INFO  [OpenTSDB I/O Worker #3] ConnectionManager: [id: 0x3518b469, /0:0:0:0:0:0:0:1:62207 => /0:0:0:0:0:0:0:1:4242] CONNECTED: /0:0:0:0:0:0:0:1:62207
2016-07-12 19:03:20,902 DEBUG [OpenTSDB I/O Worker #3] HttpContentCompressor: -Dio.netty.noJdkZlibEncoder: false
2016-07-12 19:03:21,193 DEBUG [OpenTSDB I/O Worker #3] QueryStats: New query for remote /0:0:0:0:0:0:0:1:62207 with hash 1549990544 on thread 13
2016-07-12 19:03:21,193 DEBUG [OpenTSDB I/O Worker #3] QueryStats: Successfully put new query for remote /0:0:0:0:0:0:0:1:62207 with hash 1549990544 on thread 13 w q TSQuery(start_time=1h-ago, end_time=null, subQueries[TSSubQuery(metric=interface.p2p0.if_packets.tx, filters=[filter_name=wildcard, tagk=status, group_by=true, tagk_bytes=null, config={components=[*], case=false}], tsuids=[], agg=sum, downsample=null, ds_interval=0, rate=false, rate_options=null, explicit_tags=explicit_tags, index=0), TSSubQuery(metric=interface.p2p0.if_packets.tx, filters=[filter_name=wildcard, tagk=status, group_by=true, tagk_bytes=null, config={components=[*], case=false}], tsuids=[], agg=sum, downsample=null, ds_interval=0, rate=false, rate_options=null, explicit_tags=explicit_tags, index=1), TSSubQuery(metric=interface.bridge0.if_octets.rx, filters=[filter_name=wildcard, tagk=status, group_by=true, tagk_bytes=null, config={components=[*], case=false}], tsuids=[], agg=sum, downsample=null, ds_interval=0, rate=false, rate_options=null, explicit_tags=explicit_tags, index=2)] padding=false, no_annotations=false, with_global_annotations=false, show_tsuids=false, ms_resolution=false, options=[])
2016-07-12 19:03:21,250 INFO  [OpenTSDB I/O Worker #3] QueryStats: Executing new query={"query":{"start":"1h-ago","end":null,"timezone":null,"options":null,"padding":false,"queries":[{"aggregator":"sum","metric":"interface.p2p0.if_packets.tx","tsuids":null,"downsample":null,"rate":false,"filters":[{"tagk":"status","filter":"*","group_by":true,"type":"wildcard"}],"index":0,"tags":{"status":"wildcard(*)"},"explicitTags":false,"rateOptions":null,"filterTagKs":[]},{"aggregator":"sum","metric":"interface.p2p0.if_packets.tx","tsuids":null,"downsample":null,"rate":false,"filters":[{"tagk":"status","filter":"*","group_by":true,"type":"wildcard"}],"index":1,"tags":{"status":"wildcard(*)"},"explicitTags":false,"rateOptions":null,"filterTagKs":[]},{"aggregator":"sum","metric":"interface.bridge0.if_octets.rx","tsuids":null,"downsample":null,"rate":false,"filters":[{"tagk":"status","filter":"*","group_by":true,"type":"wildcard"}],"index":2,"tags":{"status":"wildcard(*)"},"explicitTags":false,"rateOptions":null,"filterTagKs":[]}],"delete":false,"noAnnotations":false,"globalAnnotations":false,"showTSUIDs":false,"msResolution":false,"showQuery":false,"showStats":false,"showSummary":false,"useCalendar":false},"exception":"null","executed":1,"user":null,"requestHeaders":{"Origin":"chrome-extension://fdmmgilgnpjigdojojpjoooidkmcomcm","Cache-Control":"no-cache","Accept":"*/*","Connection":"keep-alive","User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36","Host":"localhost:4242","DNT":"1","Accept-Encoding":"gzip, deflate","Accept-Language":"en-US,en;q=0.8","Content-Length":"656","Content-Type":"text/plain;charset=UTF-8"},"numRunningQueries":1,"httpResponse":null,"queryStartTimestamp":1468368201193,"queryCompletedTimestamp":0,"sentToClient":false,"stats":{}}
2016-07-12 19:03:21,257 DEBUG [OpenTSDB I/O Worker #3] QueryUtil: Regex for scanner: Scanner(table="tsdb", start_key="\x00\x006W\x85vp", stop_key=[0, 0, 54, 87, -123, -110, -112], columns={"t"}, populate_blockcache=true, max_num_rows=128, max_num_kvs=4096, region=null, filter=null, scanner_id=0x0000000000000000): (?s)^.{7}(?:.{6})*\Q[0, 0, 2]E.{3}(?:.{6})*$
2016-07-12 19:03:21,270 DEBUG [OpenTSDB I/O Worker #3] QueryUtil: Regex for scanner: Scanner(table="tsdb", start_key="\x00\x006W\x85vp", stop_key=[0, 0, 54, 87, -123, -110, -112], columns={"t"}, populate_blockcache=true, max_num_rows=128, max_num_kvs=4096, region=null, filter=null, scanner_id=0x0000000000000000): (?s)^.{7}(?:.{6})*\Q[0, 0, 2]E.{3}(?:.{6})*$
2016-07-12 19:03:21,271 DEBUG [OpenTSDB I/O Worker #3] QueryUtil: Regex for scanner: Scanner(table="tsdb", start_key="\x00\x00,W\x85vp", stop_key=[0, 0, 44, 87, -123, -110, -112], columns={"t"}, populate_blockcache=true, max_num_rows=128, max_num_kvs=4096, region=null, filter=null, scanner_id=0x0000000000000000): (?s)^.{7}(?:.{6})*\Q[0, 0, 2]E.{3}(?:.{6})*$
2016-07-12 19:03:21,279 INFO  [AsyncHBase I/O Worker #1] TsdbQuery: TsdbQuery(start_time=1468364601192, end_time=1468368201192, metric=[0, 0, 54], filters=[filter_name=wildcard, tagk=status, group_by=true, tagk_bytes=[0, 0, 2], config={components=[*], case=false}], rate=false, aggregator=sum, group_bys=(status [0, 0, 2])) matched 1 rows in 1 spans in 17.436322ms
2016-07-12 19:03:21,281 INFO  [AsyncHBase I/O Worker #1] TsdbQuery: TsdbQuery(start_time=1468364601192, end_time=1468368201192, metric=[0, 0, 54], filters=[filter_name=wildcard, tagk=status, group_by=true, tagk_bytes=[0, 0, 2], config={components=[*], case=false}], rate=false, aggregator=sum, group_bys=(status [0, 0, 2])) matched 1 rows in 1 spans in 10.633964ms
2016-07-12 19:03:21,281 INFO  [AsyncHBase I/O Worker #1] TsdbQuery: TsdbQuery(start_time=1468364601192, end_time=1468368201192, metric=[0, 0, 44], filters=[filter_name=wildcard, tagk=status, group_by=true, tagk_bytes=[0, 0, 2], config={components=[*], case=false}], rate=false, aggregator=sum, group_bys=(status [0, 0, 2])) matched 1 rows in 1 spans in 10.462772ms
2016-07-12 19:03:21,284 DEBUG [AsyncHBase I/O Worker #1] AggregationIterator: Aggregating 1 iterators
2016-07-12 19:03:21,288 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Added results for m1 to e1
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] AggregationIterator: Aggregating 1 iterators
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Added results for m2 to e2
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] AggregationIterator: Aggregating 1 iterators
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Added results for m1 to e1
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] AggregationIterator: Aggregating 1 iterators
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Added results for m2 to e2
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] AggregationIterator: Aggregating 1 iterators
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Added results for m3 to e3
2016-07-12 19:03:21,289 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Nested expression detected. e12 depends on e1
2016-07-12 19:03:21,308 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Nested expression detected. e12 depends on e2
2016-07-12 19:03:21,316 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Expressions Size is 4
2016-07-12 19:03:21,316 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Adding expression to compile_stack[0]
2016-07-12 19:03:21,317 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Adding expression to compile_stack[1]
2016-07-12 19:03:21,318 DEBUG [AsyncHBase I/O Worker #1] QueryExecutor: Adding expression to compile_stack[2]
2016-07-12 19:03:21,322 ERROR [AsyncHBase I/O Worker #1] QueryExecutor: Query exception:
java.io.IOException:  Internal Error: Less expressions where added to the compile stack than expressions.size (3 instead of 4)
	at net.opentsdb.tsd.QueryExecutor$1QueriesCB.call(QueryExecutor.java:330) ~[tsdb-2.3.0-RC1.jar:3dc2553]
	at net.opentsdb.tsd.QueryExecutor$1QueriesCB.call(QueryExecutor.java:252) ~[tsdb-2.3.0-RC1.jar:3dc2553]
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1278) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1257) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.callback(Deferred.java:1005) [async-1.4.0.jar:na]
	at com.stumbleupon.async.DeferredGroup.done(DeferredGroup.java:173) [async-1.4.0.jar:na]
	at com.stumbleupon.async.DeferredGroup.recordCompletion(DeferredGroup.java:158) [async-1.4.0.jar:na]
	at com.stumbleupon.async.DeferredGroup.access$200(DeferredGroup.java:36) [async-1.4.0.jar:na]
	at com.stumbleupon.async.DeferredGroup$1NotifyOrdered.call(DeferredGroup.java:97) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1278) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1257) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.callback(Deferred.java:1005) [async-1.4.0.jar:na]
	at net.opentsdb.core.TsdbQuery$1ScannerCB.close(TsdbQuery.java:873) [tsdb-2.3.0-RC1.jar:3dc2553]
	at net.opentsdb.core.TsdbQuery$1ScannerCB.call(TsdbQuery.java:633) [tsdb-2.3.0-RC1.jar:3dc2553]
	at net.opentsdb.core.TsdbQuery$1ScannerCB.call(TsdbQuery.java:575) [tsdb-2.3.0-RC1.jar:3dc2553]
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1278) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1257) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.access$300(Deferred.java:430) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred$Continue.call(Deferred.java:1366) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1278) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1257) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.callback(Deferred.java:1005) [async-1.4.0.jar:na]
	at org.hbase.async.HBaseRpc.callback(HBaseRpc.java:698) [asynchbase-1.7.1.jar:na]
	at org.hbase.async.RegionClient.decode(RegionClient.java:1516) [asynchbase-1.7.1.jar:na]
	at org.hbase.async.RegionClient.decode(RegionClient.java:88) [asynchbase-1.7.1.jar:na]
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:500) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [netty-3.9.4.Final.jar:na]
	at org.hbase.async.RegionClient.handleUpstream(RegionClient.java:1206) [asynchbase-1.7.1.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.SimpleChannelHandler.messageReceived(SimpleChannelHandler.java:142) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.handler.timeout.IdleStateAwareChannelHandler.handleUpstream(IdleStateAwareChannelHandler.java:36) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.handler.timeout.IdleStateHandler.messageReceived(IdleStateHandler.java:294) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559) [netty-3.9.4.Final.jar:na]
	at org.hbase.async.HBaseClient$RegionClientPipeline.sendUpstream(HBaseClient.java:3108) [asynchbase-1.7.1.jar:na]
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42) [netty-3.9.4.Final.jar:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_92]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_92]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_92]
2016-07-12 19:03:21,324 ERROR [AsyncHBase I/O Worker #1] HttpQuery: [id: 0x3518b469, /0:0:0:0:0:0:0:1:62207 => /0:0:0:0:0:0:0:1:4242] Internal Server Error on /api/query/exp
java.io.IOException:  Internal Error: Less expressions where added to the compile stack than expressions.size (3 instead of 4)
	at net.opentsdb.tsd.QueryExecutor$1QueriesCB.call(QueryExecutor.java:330) ~[tsdb-2.3.0-RC1.jar:3dc2553]
	at net.opentsdb.tsd.QueryExecutor$1QueriesCB.call(QueryExecutor.java:252) ~[tsdb-2.3.0-RC1.jar:3dc2553]
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1278) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1257) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.callback(Deferred.java:1005) [async-1.4.0.jar:na]
	at com.stumbleupon.async.DeferredGroup.done(DeferredGroup.java:173) [async-1.4.0.jar:na]
	at com.stumbleupon.async.DeferredGroup.recordCompletion(DeferredGroup.java:158) [async-1.4.0.jar:na]
	at com.stumbleupon.async.DeferredGroup.access$200(DeferredGroup.java:36) [async-1.4.0.jar:na]
	at com.stumbleupon.async.DeferredGroup$1NotifyOrdered.call(DeferredGroup.java:97) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1278) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1257) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.callback(Deferred.java:1005) [async-1.4.0.jar:na]
	at net.opentsdb.core.TsdbQuery$1ScannerCB.close(TsdbQuery.java:873) [tsdb-2.3.0-RC1.jar:3dc2553]
	at net.opentsdb.core.TsdbQuery$1ScannerCB.call(TsdbQuery.java:633) [tsdb-2.3.0-RC1.jar:3dc2553]
	at net.opentsdb.core.TsdbQuery$1ScannerCB.call(TsdbQuery.java:575) [tsdb-2.3.0-RC1.jar:3dc2553]
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1278) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1257) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.access$300(Deferred.java:430) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred$Continue.call(Deferred.java:1366) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.doCall(Deferred.java:1278) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.runCallbacks(Deferred.java:1257) [async-1.4.0.jar:na]
	at com.stumbleupon.async.Deferred.callback(Deferred.java:1005) [async-1.4.0.jar:na]
	at org.hbase.async.HBaseRpc.callback(HBaseRpc.java:698) [asynchbase-1.7.1.jar:na]
	at org.hbase.async.RegionClient.decode(RegionClient.java:1516) [asynchbase-1.7.1.jar:na]
	at org.hbase.async.RegionClient.decode(RegionClient.java:88) [asynchbase-1.7.1.jar:na]
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:500) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [netty-3.9.4.Final.jar:na]
	at org.hbase.async.RegionClient.handleUpstream(RegionClient.java:1206) [asynchbase-1.7.1.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.SimpleChannelHandler.messageReceived(SimpleChannelHandler.java:142) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.handler.timeout.IdleStateAwareChannelHandler.handleUpstream(IdleStateAwareChannelHandler.java:36) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.handler.timeout.IdleStateHandler.messageReceived(IdleStateHandler.java:294) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559) [netty-3.9.4.Final.jar:na]
	at org.hbase.async.HBaseClient$RegionClientPipeline.sendUpstream(HBaseClient.java:3108) [asynchbase-1.7.1.jar:na]
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108) [netty-3.9.4.Final.jar:na]
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42) [netty-3.9.4.Final.jar:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_92]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_92]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_92]
2016-07-12 19:03:21,335 INFO  [AsyncHBase I/O Worker #1] HttpQuery: [id: 0x3518b469, /0:0:0:0:0:0:0:1:62207 => /0:0:0:0:0:0:0:1:4242] HTTP /api/query/exp done in 422ms```

**Comment by johann8384** [comment]
I found the cause.

If there are no nested queries, graph is null, it doesn't build one, and goes here: (line 389 or so)

} else {
  for (final ExpressionIterator ei : expressions.values()) {
    ei.compile();
    LOG.debug("Successfully compiled " + ei);
  }
} 

If there are nested queries, it uses the graph:

if (graph != null) {

  final Integer expressionLength = expressions.size();
  final ExpressionIterator[] compile_stack = new ExpressionIterator[expressionLength];
  final TopologicalOrderIterator<String, DefaultEdge> it = new TopologicalOrderIterator<String, DefaultEdge>(graph);
The problem is when you have an expression that is not a part of a nested query, so if you have e1, e2, e3 and e12, e12 uses e1 and e3 but not e2. e2 will never be added to the graph and then exists as a null ExpressionIterator.
